fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/JobsServlet.java

```diff
@@ -359,20 +355,50 @@ protected void handlePreResponse(
      * notification before timeout, we retrieve the PreResponse from the PreResponseStore else we return an empty
      * Observable.
      *
-     * @param ticket  The ticket whose associated PreResponse needs to be retrieved.
-     * @param apiRequest  JobsApiRequest object with all the associated info in it
+     * @param ticket  The ticket for which the PreResponse needs to be retrieved.
+     * @param asyncAfter  The minimum duration the request is allowed to last before becoming asynchronous
+     *
+     * @return An Observable wrapping a PreResponse or an empty Observable in case a timeout occurs.
+     */
+    protected Observable<PreResponse> getResults(@NotNull String ticket, long asyncAfter) {
+        /*
+         * BroadCastChannel is a hot observable i.e. it emits notification irrespective of whether it has any
+         * subscribers. We use the replay operator so that the preResponseObservable upon connection, will begin
+         * collecting values.
+         * Once a new observer subscribes to the observable, it will have all the collected values replayed to it.
+         */
+        if (asyncAfter == JobsApiRequest.ASYNCHRONOUS_ASYNC_AFTER_VALUE) {
+            return Observable.empty();
+        } else {
+            ConnectableObservable<String> broadcastChannelNotifications = broadcastChannel.getNotifications()
+                    .filter(ticket::equals)
+                    .take(1)
+                    .replay(1);
+            broadcastChannelNotifications.connect();
+            return preResponseStore.get(ticket).switchIfEmpty(
+                    applyTimeoutIfNeeded(broadcastChannelNotifications, asyncAfter).flatMap(preResponseStore::get)
```

Comment:
Are you sure this is doing the correct thing? To me, this looks like it will wait until `preResponseStore.get()` completes, and then activate the "timeout" (if needed).

I'm _guessing_ this is now doing the right thing, but it's not immediately obvious, so more comments are likely needed around what's happening, and why it's correct. (talking about the ordering of things might not hurt either)